### DISCUSSION TITLE
forwardRef

### DISCUSSION THREAD

**Discussion by honaseung** [honaseung/react-deep-dive#27] [Chapter3]
You know the `ref`. Right? We talked about it few weeks ago. We can use `ref` for their own `HTML` it's not all purpose of `ref` but usually does. Then what about child component's `HTML`? I mean child also has theirs right? So it's ok to save child component's `HTML` as a `ref`.

When you try give your `ref` variable as a props of child component, it doesn't work well. Because `ref` is reserved you can't use it freely. Then how can we do? Actually it's simple. just don't use`ref`. You can just use like `somethingRef`. That's all. Now it works well. But it's quite few hard to make a name.(Honestly, I don't agreed with this idea.) So `forwardRef` is born. You can use `ref` with forwardRef.
```
const P = () => {
  const cRef = useRef(null);
  .
  .
  .
  return (
    <C ref={cRef} />
  );
}
```
```
const C = forwardRef((props, ref) => {
  .
  .
  .
  return (
    <input ref={ref} />
  );
});
```
Now you can see child component's `HTML`.

**Comment by honaseung** [comment]
`ref` 에 대해서는 아실겁니다. `ref` 를 사용하는 이유의 전부는 아니지만 우리는 주로 컴포넌트들의 `HTML` 를 저장하는 용도로 사용합니다. 그렇다면 자식 컴포넌트들의 `HTML` 은 어떨까요? 자식들도 `HTML` 을 가지고 있기 때문에 그것들을 저장하는 것은 괜찮습니다.

당신이 `ref` 변수를 자식 컴포넌트의 프롭스로 주려고 한다면 작동하지 않을 것입니다. 왜냐하면 `ref` 는 당신이 자유롭게 사용할 수 없는 예약어이기 때문입니다. 그렇다면 어떻게 해야할까요? 사실 간단합니다. 그냥 `ref` 만 안쓰면 돼요. 당신은 단순하게 `somethingRef` 라고도 쓸 수 있습니다. 이젠 정상작동 할 것입니다. 이름을 짓는건 어려운 일이죠. (솔직히 동의하진 않습니다만) 그래서 `forwardRef` 가 탄생했다고 합니다. 당신은 `forwardRef` 와 함께 일때 `ref` 를 사용할 수 있습니다.
```
const P = () => {
  const cRef = useRef(null);
  .
  .
  .
  return (
    <C ref={cRef} />
  );
}
```
```
const C = forwardRef((props, ref) => {
  .
  .
  .
  return (
    <input ref={ref} />
  );
});
```

**Reply by chaminju927** [reply]
ref변수를 자식 컴포넌트에게 props로 내려주려 할때, 예약어로 ref만 사용하지 않으면 된다. <- 이 부분이 잘 이해가 가지 않는데 더 자세하게 설명해주실수 있나요?

**Reply by honaseung** [reply]
만약 자식에게 `ref` 값을 `ref={someRef}` 라고 직접 `ref` 에 전달하려 한다면 에러가 발생하게 된다는 의미입니다.
